CL-1825 Standardize tables

luucvanderzee opened the pull request:
## Notes

Still a few API-related things we should discuss:

- I ended up going for the `chakra-ui` component names. The `semantic-ui` ones are like `Row` and `Header` and there are already way too many things with those names. `Thead` and `Tr` avoid variable name conflicts and more explicitly say that they're used in the context of tables
- I am not sure about putting those `innerBorders` on the `Table` component. Very often you create a `Table` and then create seperate wrapper components for the `Body` and `Rows` etc. If you're then trying to figure out where the borders are coming from it's a bit unintuitive to me that it's from the `Table` and not e.g. the `Body`. So I would be in favor of moving these props to `Thead` and `Tbody`
- I think adding a `background` to the `Tr` component would be useful. I notice that I am having to add backgrounds to all `Td`s in a single row when the row is selected / hovered etc. In this case we could also remove these backgrounds from the `Thead` and `TFoot`
- Some tables didn't have the `border-bottom` between the header and the body. I didn't want to make another border flag so now these tables just have an extra border lol (see e.g. users table)

## Tables changed

- Representativeness dashboard: table view (not enabled on epic, maybe because no custom sign up question)
- Insights input table (not enabled on epic)
- [Activities table](https://1825.epic.citizenlab.co/en/admin/moderation)
- [Users table](https://1825.epic.citizenlab.co/en/admin/users)
- `PostManager` tables, e.g. [input manager](https://1825.epic.citizenlab.co/en/admin/ideas)
- Invitations table (not visible on epic because there are no invitations I guess)

## Checklist

- [x] Added entry to changelog
- [x] Prepared branch for code review

## Links

- [Epic Deployment](https://1825.epic.citizenlab.co/)

## How urgent is a code review?

I can work with this for now so not extremely urgent on my side

luucvanderzee committed:
[CL-1744] Rebuilding semantic UI table (WIP)

luucvanderzee committed:
Merge branch 'master' into CL-1744-add-referrers-table

luucvanderzee committed:
[CL-1744] Create constants file

luucvanderzee committed:
[CL-1744] Add Row and Cell

luucvanderzee committed:
[CL-1744] Creating dummy table for testing (WIP)

luucvanderzee committed:
[CL-1744] Add header cell

luucvanderzee committed:
[CL-1744] _

luucvanderzee committed:
[CL-1744] Fix corner radius and add box props to Cell and HeaderCell

luucvanderzee committed:
[CL-1744] Add box props to Table

luucvanderzee committed:
[CL-1744] Add box pprops to all table components

luucvanderzee committed:
[CL-1744] Set up borders correctly

luucvanderzee committed:
[CL-1744] Remove box props from components where they don't make sense

luucvanderzee committed:
[CL-1744] Remove Footer

luucvanderzee committed:
[CL-1744] Representation: hide border top if necessary

luucvanderzee committed:
[CL-1744] Make first column in representativeness table same color as header

luucvanderzee committed:
[CL-1744] _

luucvanderzee committed:
[CL-1744] Add vertical header borders

luucvanderzee committed:
[CL-1744] Add clickable prop to HeaderCell

luucvanderzee committed:
[CL-1744] _

luucvanderzee committed:
[CL-1744] Replace semantic ui tabe by new table in representativeness

luucvanderzee committed:
[CL-1744] Clean up invitations Row

luucvanderzee committed:
[CL-1744] Refactor invites table to functional component

luucvanderzee committed:
[CL-1744] Extract table header

luucvanderzee committed:
[CL-1744] Replace semantic-ui Popup by IconTooltip

luucvanderzee committed:
[CL-1744] Add Footer component

luucvanderzee committed:
[CL-1744] Replace semantic ui table by new table

luucvanderzee committed:
[CL-1744] Fix footer

luucvanderzee committed:
Fix hardcoded page numbers

luucvanderzee committed:
Fix unit tests

luucvanderzee committed:
Merge branch 'master' into replace-semantic-ui-tables

luucvanderzee committed:
[CL-1825] Refactor Table API

luucvanderzee committed:
[CL-1825] Refactor representativeness table with new API

luucvanderzee committed:
[CL-1825] Also do invitations

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Fix double border issue

luucvanderzee committed:
[CL-1825] Replace semantic-ui colors

luucvanderzee committed:
[CL-1825] Remove clickable prop from HeaderCell

luucvanderzee committed:
[CL-1825] Remove default Header and Footer background again

luucvanderzee committed:
[CL-1825] Add sort logic to header cell after all

luucvanderzee committed:
[CL-1825] Replace table in PostManager (WIP)

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Fix border issue

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Fix drag and drop

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Ideas table: fix sorting behavior

luucvanderzee committed:
[CL-1825] Remove TableHeaderCellText

luucvanderzee committed:
[CL-1825] Have default info tooltip

luucvanderzee committed:
[CL-1825] Remove sortability of 'Picks'

luucvanderzee committed:
[CL-1825] Change table font size

luucvanderzee committed:
[CL-1825] Move all row related stuff in Row folder

luucvanderzee committed:
[CL-1825] Fix idea text color

luucvanderzee committed:
[CL-1825] Add active backround stuff

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Force upvotes and downvotes content to stay on one line

luucvanderzee committed:
[CL-1825] Fix column widths

luucvanderzee committed:
[CL-1825] Improve idea header row typing

luucvanderzee committed:
[CL-1825] Fix initiatives table header

luucvanderzee committed:
[CL-1825] Fix StyledRow situation

luucvanderzee committed:
[CL-1825] Initiatives table: fix cell background if active

luucvanderzee committed:
[CL-1825] Rename table components to follow chakra conventions

luucvanderzee committed:
[CL-1825] Replace customizable inner borders by booleans for now

luucvanderzee committed:
[CL-1825] PostManager: also highlight sorted column

luucvanderzee committed:
[CL-1825] Fix header and add body row borders

luucvanderzee committed:
[CL-1825] Set column names to uppercase

luucvanderzee committed:
[CL-1825] Fix spacing

luucvanderzee committed:
[CL-1825] Rename: Capitalize -> Uppercase

luucvanderzee committed:
[CL-1825] Fix icon size

luucvanderzee committed:
[CL-1825] Fix URl behavior

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Users table: fix sorting + upper case column names

luucvanderzee committed:
[CL-1825] Add background select color

luucvanderzee committed:
[CL-1825] Insights table: refactor with new table components

luucvanderzee committed:
[CL-1825] Fix view if query active

luucvanderzee committed:
[CL-1825] Use sortable th

luucvanderzee committed:
[CL-1825] Remove unused components/UI/Table

luucvanderzee committed:
[CL-1825] CHANGELOG

luucvanderzee committed:
Merge branch 'master' into CL-1825-standardize-tables

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Delete SortableTableHeader

luucvanderzee committed:
[CL-1825] Fix linter errors

luucvanderzee committed:
[CL-1825] Fix small sort icon and tooltip icon not staying on same line

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Fix unit tests

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/Row/InitiativeRow.tsx:
```diff
@@ -16,11 +16,14 @@ import {
+import { TitleLink } from '.';
+import StyledRow from './StyledRow';
+import { Td } from 'components/admin/Table';
+import { Icon } from 'semantic-ui-react';
```
Would be nice to get rid of these icons, too, but not blocking.

luucvanderzee replied:
Definitely, but they're still used all over the place unfortunately. Would be better to do them in one go in collaboration with the design team, so that they can pick which `semantic-ui` icons we replace with our own icons etc.

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/Row/InitiativeRow.tsx:
```diff
@@ -72,6 +75,22 @@ interface InputProps {
+
+const Cell = ({ onClick, active, children }: CellProps) => (
+  <Td
+    borderBottom="none !important"
```
Hm, not a fan of this, perhaps we should find a way to avoid it.

luucvanderzee replied:
The `borderBottom` is coming from the `Table`'s `innerBorders.bodyRows` prop. It uses a selector that has a higher specificity. I wasn't able to figure out how to override it without `!important`... an alternative we could consider is adding `!important` by default inside of the `Td` component for the `borderBottom` prop?

IvaKop replied:
Let's see how many use cases we'll have for this and we can amend it, if needed. If this is the only place for now, I'd say it's fine.

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/Row/StyledRow.tsx:
```diff
@@ -0,0 +1,22 @@
+import { Props as RowProps } from 'components/admin/Table/Tr';
+
+const StyledRow = styled(Tr)<{ undraggable: boolean }>`
+  height: 5.7rem !important;
```
Why do we need an important attribute here? Shouldn't it "just work" without it?

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/Row/StyledRow.tsx:
```diff
@@ -0,0 +1,22 @@
+
+import { Props as RowProps } from 'components/admin/Table/Tr';
+
+const StyledRow = styled(Tr)<{ undraggable: boolean }>`
```
This type should also include the RowProps, I think..

luucvanderzee replied:
I believe `styled-components` automatically forwards all other props so this type argument is only for your own sanity if you want to use any of the props inside of the style definition. Since we're not using the other types I think it's ok like this

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/header/IdeaHeaderRow.tsx:
```diff
@@ -0,0 +1,254 @@
+import { SortDirection } from 'utils/paginationUtils';
+
+interface SortableHeaderCellProps {
+  sortAttribute?: IdeasSortAttribute;
```
All these sort props should be grouped, I believe but if it was like that in the past already, could be out of scope for this PR.

luucvanderzee replied:
Also agree but indeed I just copied what was already there. We're using this pattern in a bunch of different places, might be good to replace all of them in one go

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/header/InitiativesHeaderRow.tsx:
```diff
@@ -0,0 +1,78 @@
+
+// components
+import { Thead, Tr, Th } from 'components/admin/Table';
+import Checkbox from 'components/UI/Checkbox';
```
Could we use the Checkbox from the component library? I want to delete this one at some point..

IvaKop commented on front/app/components/admin/PostManager/components/PostTable/index.tsx:
```diff
@@ -78,8 +79,8 @@ const Container = styled.div`
 
-export const TableHeaderCellText = styled.span`
-  font-weight: 600;
+const StyledTable = styled(Table)`
```
Really needed? Could we use our base font size?

luucvanderzee replied:
Yeah screw it

IvaKop commented on front/app/components/admin/Table/Td.tsx:
```diff
@@ -0,0 +1,16 @@
+}
+
+const Cell = ({ children, colSpan, ...otherProps }: Props) => (
+  <Box as="td" p="12px" colSpan={colSpan as any} {...otherProps}>
```
Why is `as any` needed here? You can use a styled box or just pass this in a `style` object for convenience?

luucvanderzee replied:
Turns out `colSpan` is a `number` in React, changed it and now the `any` is not necessary anymore :)

IvaKop commented on front/app/components/admin/Table/Th.tsx:
```diff
@@ -0,0 +1,70 @@
+  infoTooltip,
+  ...otherProps
+}: Props) => (
+  <StyledBox as="th" p="12px" colSpan={colSpan as any} {...otherProps}>
```
Also here, not sure why this casting is necessary.

IvaKop commented on front/app/components/admin/Table/Th.tsx:
```diff
@@ -0,0 +1,70 @@
+  <StyledBox as="th" p="12px" colSpan={colSpan as any} {...otherProps}>
+    {children}
+
+    {sortDirection && (
```
Love the prop structure for this - one prop determines the entire behaviour 👍

IvaKop commented on front/app/components/admin/Table/Thead.tsx:
```diff
@@ -0,0 +1,19 @@
+}
+
+const StyledTHead = styled.thead<{ background?: string }>`
+  background: ${({ background }) => background ?? 'white'};
```
Could you use `colors.white` - not a practical different atm but for consistency, it's better.

IvaKop commented on front/app/components/admin/Table/Tr.tsx:
```diff
@@ -0,0 +1,19 @@
+const Row = React.forwardRef(
+  (
+    { children, className }: Props,
+    ref: React.RefObject<HTMLTableRowElement>
```
This is needed to support re-ordering, I guess?

luucvanderzee replied:
Yeah, although I think we could also just rename `ref` to `innerRef` or something and avoid this `forwardRef` stuff... what do you think?

IvaKop replied:
I think it's fine to leave it as is, tbh. No need to introduce custom props.

IvaKop commented on front/app/components/admin/Table/index.tsx:
```diff
@@ -0,0 +1,7 @@
+export { default as Table } from './Table';
```
Love this, very clear!

IvaKop commented on front/app/containers/Admin/users/UserTableRow.tsx:
```diff
@@ -190,28 +191,32 @@ class UserTableRow extends PureComponent<Props & InjectedIntlProps, State> {
         className={`e2e-user-table-row ${selected ? 'selected' : ''}`}
       >
-        <td>
+        <Td background={selected ? colors.background : ''}>
```
Should it be `: undefined`?

IvaKop commented on front/app/modules/commercial/moderation/admin/containers/ModerationRow.tsx:
```diff
@@ -72,8 +73,8 @@ const GoToLink = styled(Link)`
 const GoToIcon = styled(Icon)`
-  width: 100%;
-  height: 100%;
+  width: 14px;
```
This will probably be too small with the new icons. Can we leave it without a custom size?

luucvanderzee replied:
Ben also said this was too small, fixed it :+1:

IvaKop reviewed (commented):
I have some really minor comments throughout the PR, none of which are really blocking, I think. 

Just one question - I think we need some unit tests for these new components, especially if we plan to re-use them all over the place. Do you think we can add them?

On the questions you raised:

> I ended up going for the chakra-ui component names. The semantic-ui ones are like Row and Header and there are already way too many things with those names. Thead and Tr avoid variable name conflicts and more explicitly say that they're used in the context of tables

This is great, it makes a lot of sense to me!

> I am not sure about putting those innerBorders on the Table component. Very often you create a Table and then create seperate wrapper components for the Body and Rows etc. If you're then trying to figure out where the borders are coming from it's a bit unintuitive to me that it's from the Table and not e.g. the Body. So I would be in favor of moving these props to Thead and Tbody

I still feel it makes sense to configure the borders in a single place for the table and not spread this configuration across multiple components. While I agree it could potentially be a source of confusion, I think we can easily remedy this by communicating with the team and documenting the component properly in Storybook / Notion.

> I think adding a background to the Tr component would be useful. I notice that I am having to add backgrounds to all Tds in a single row when the row is selected / hovered etc. In this case we could also remove these backgrounds from the Thead and TFoot

I agree - it's a good idea.

> Some tables didn't have the border-bottom between the header and the body. I didn't want to make another border flag so now these tables just have an extra border lol (see e.g. users table)

I think that's totally fine, unless the design team is strongly against it :)

luucvanderzee committed:
Merge branch 'master' into CL-1825-standardize-tables

luucvanderzee committed:
[CL-1825] Fix icon size in activities table

luucvanderzee committed:
[CL-1825] Remove impotant from post table row height

luucvanderzee committed:
[CL-1825] Various small things

luucvanderzee committed:
[CL-1825] Change last two colSpans

luucvanderzee committed:
[CL-1825] Remove background from Thead and Tfoot, put on Tr instead

luucvanderzee committed:
[CL-1825] Move Td background to Tr background where possible

luucvanderzee committed:
[CL-1825] _

luucvanderzee committed:
[CL-1825] Replace UI/Checkbox by component library checkbox

luucvanderzee commented:
I removed `background` from the `Thead` and `Tfoot` and put it on the `Tr` instead. Also removed the `background`s from `Td` and replaced it by a single background on the `Tr` where it made sense. Left some comments on your comments and resolved the other things!

luucvanderzee committed:
[CL-1825] Fix linter error

luucvanderzee commented:
As for the unit tests to be honest it's all pretty basic... what kind of key functionality do you think we should test?

IvaKop commented:
@luucvanderzee Unit tests - for the basic components - just that they render correctly with some content. For the sorting and info icon functionality - that they show what they should with the appropriate props. It's a sanity check, essentially, nothing complex.

IvaKop reviewed (approved):
Looks good - let's just add the unit tests and we are good to go + move it to the component library 🚀

luucvanderzee committed:
Merge branch 'master' into CL-1825-standardize-tables

luucvanderzee committed:
[CL-1825] Upgrade component library

luucvanderzee committed:
[CL-1825] Delete Table and use Table stuff from component library instead

luucvanderzee committed:
[CL-1825] Remove duplicate imports
